src/libcrun/kontain_nonkmexec.c

```diff
@@ -0,0 +1,343 @@
+/*
+ * crun - OCI runtime written in C
+ *
+ * Copyright (C) 2020 Kontain Inc.
+ * crun is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation; either version 2.1 of the License, or
+ * (at your option) any later version.
+ *
+ * crun is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with crun.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <stdio.h>
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <unistd.h>
+#include <fcntl.h>
+#include <pthread.h>
+#include <string.h>
+#include <time.h>
+#include <sys/queue.h>
+#include <regex.h>
+#include <stdlib.h>
+#include <errno.h>
+#include "utils.h"
+#include "sha256.h"
+
+/*
+ * What is going on in here?
+ * We need to allow non kontain programs to be run inside of a kontain container.
+ * But, we don't want arbitrary programs to be run.  So, we have a list of acceptable
+ * programs that "docker exec ...." should be able to run.  This list is stored in
+ * file (see KONTAIN_KRUN_CONFIG definition below).  When it is time to do the
+ * exec() system call we parse the file, check to see if the program we are execing
+ * too is allowed to run without km.  If it is allowed we exec the program directly.
+ * If not, the program will need to run within a km virtual machine.
```

Comment:
I would simply say "since it circumvents KM VM encapsulation, to allow that an admin would need to provide a list of allowed exceptions in a config file (see KONTAIN_KRUN_CONFIG). Each line in the file has a regexp, the full path to an executable and a SHA - when a name passed to krun marches the regexp, krun will instead run the executable from the path. SHA is used to validate the exectubale signature to provide extra layer of security"

BTW, I found this mapping confusing. Maybe an example ? e.g. 
"/bin/ping /bin/ping XXX" is a valid line, right ? (assuming XXX is a SHA :-) ?

Reply:
The description of the line format is by the function that examines the config file (kontain_execpath_builddb()).
I find the description is more useful near the code that is doing the work.

This comment block at the beginning of the source file is to help the reader understand generally what is in the source file.
I will add a comment near the function I mentioned.